### ISSUE TITLE
[Feature] Dynamic offset text

### ISSUE THREAD

**Issue by Akianonymus** [akinsho/nvim-bufferline.lua#166]
Currently, the offset text at top is static, fetched during initialisation and then it uses the same.

How about using a dynamic text ? Adding something like offset_type option, which can be string or function

If offset_type is string, then show as it is, otherwise evaluate the string ( using loadstring ) and use it.

**Comment by siduck76** [comment]
I think it already does 

`local current_cwd = vim.fn.fnamemodify(vim.fn.fnamemodify.getcwd(), ':t');

 text = current_cwd
`

**Comment by Akianonymus** [comment]
@siduck76 No it doesn't, obviously i checked with a function before creating this issue.

It won't because it doesn't evaluate it further, it gets what it gets at the first time.

Also, the function i was using.

```
function nvim_tree_cwd()
    local cwd = vim.fn.getcwd()
    local tree_loaded, lib = pcall(require, "nvim-tree.lib")
    if tree_loaded then
        cwd = lib.Tree.cwd
    end
    local root_folder_modifier = vim.g.nvim_tree_root_folder_modifier or ":t"
    cwd = vim.fn.fnamemodify(cwd, root_folder_modifier)
    local final_cwd = cwd

    if cwd and cwd ~= "/" then
        if tree_loaded then
            local utils = require "nvim-tree.utils"
            final_cwd = utils.path_join(utils.path_remove_trailing(cwd), "..")
        end
    end
    return final_cwd
end
```

**Comment by siduck76** [comment]
> @siduck76 No it doesn't, obviously i checked with a function before creating this issue.
> 
> It won't because it doesn't evaluate it further, it gets what it gets at the first time.
> 
> Also, the function i was using.
> 
> ```
> function nvim_tree_cwd()
>     local cwd = vim.fn.getcwd()
>     local tree_loaded, lib = pcall(require, "nvim-tree.lib")
>     if tree_loaded then
>         cwd = lib.Tree.cwd
>     end
>     local root_folder_modifier = vim.g.nvim_tree_root_folder_modifier or ":t"
>     cwd = vim.fn.fnamemodify(cwd, root_folder_modifier)
>     local final_cwd = cwd
> 
>     if cwd and cwd ~= "/" then
>         if tree_loaded then
>             local utils = require "nvim-tree.utils"
>             final_cwd = utils.path_join(utils.path_remove_trailing(cwd), "..")
>         end
>     end
>     return final_cwd
> end
> ```
> 
> 

Weird this worked for me before , now it says *local undefined*

**Comment by akinsho** [comment]
@Akianonymus, so the offsets are checked and redrawn each time the bufferline reloads, but the value pulled in from the config is only done once on startup

I'm open to this, generally I think the potential solution doesn't need to be as complex as using `loadstring` or anything like that. Just changing the type signature of `text` to also take a function and doing a `type(text) == 'function' then text = text() end` or something. I guess a few things to note is that this will only be run when the tabline naturally refreshes, which is quite often anyway. I'm not open to any change that tries to increase the amount of time it refreshes or ties its refreshing to some event (a user can already do this themselves)

Anyway, I'd be fine with a PR for this and happy to help guide it. I won't implement this myself, since I don't really have a use case for it personally.

**Comment by Akianonymus** [comment]
> @Akianonymus, so the offsets are checked and redrawn each time the bufferline reloads, but the value pulled in from the config is only done once on startup
> 
> I'm open to this, generally I think the potential solution doesn't need to be as complex as using `loadstring` or anything like that. Just changing the type signature of `text` to also take a function and doing a `type(text) == 'function' then text = text() end` or something. I guess a few things to note is that this will only be run when the tabline naturally refreshes, which is quite often anyway. I'm not open to any change that tries to increase the amount of time it refreshes or ties its refreshing to some event (a user can already do this themselves)
> 
> Anyway, I'd be fine with a PR for this and happy to help guide it. I won't implement this myself, since I don't really have a use case for it personally.

hmm, thx for the idea, i am new to lua so didn't it can be done without loadstring too.

Anyway, i have created the pr: https://github.com/akinsho/nvim-bufferline.lua/pull/170

About the documentations, i would be glad if you could handle that or atleast hint what to add.

**Comment by akinsho** [comment]
closed by #170
